File: Minimal/tests/Brontide.Minimal.Host.Tests/ComponentProviderTrustOfflinePolicyTests.fs
namespace Brontide.Minimal.Host.Tests

open System
open System.IO
open System.Text.Json
open Brontide.Minimal.Host
open NUnit.Framework

type private Cbi49TemporaryJournal() =
    let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi49-{Guid.NewGuid():N}")
    member _.Path = Path.Combine(root, "cadence.bin")
    interface IDisposable with
        member _.Dispose() =
            try if Directory.Exists root then Directory.Delete(root, true)
            with :? IOException | :? UnauthorizedAccessException -> ()

[<TestFixture>]
type ComponentProviderTrustOfflinePolicyTests() =
    let runIdentity = ProviderTrustCadenceRunId.create "cadence-run.cbi49"
    let start = DateTimeOffset(2026, 8, 5, 8, 0, 0, TimeSpan.Zero)
    let schedule = ProviderServingTrustCadenceSchedule.create 2 (TimeSpan.FromMinutes 1.0)

    let interrupted path =
        let journal =
            DurableProviderTrustCadenceJournal.Establish(path, runIdentity, schedule, start).Journal.Value
        Assert.That(journal.BeginCycle().Code, Is.EqualTo "durable-cadence-cycle-started")
        DurableProviderTrustCadenceJournal.Open(path, runIdentity).Journal.Value

    let policy () = ProviderTrustOfflinePolicy.create (TimeSpan.FromMinutes 5.0) (TimeSpan.FromMinutes 1.0)

    [<Test>]
    member _.``CBI49 C1 offline policy is explicit and bounded``() =
        Assert.Throws<ArgumentException>(Action(fun () ->
            ProviderTrustOfflinePolicy.create TimeSpan.Zero (TimeSpan.FromMinutes 1.0) |> ignore)) |> ignore
        Assert.Throws<ArgumentException>(Action(fun () ->
            ProviderTrustOfflinePolicy.create (TimeSpan.FromMinutes 1.0) (TimeSpan.FromMinutes 2.0) |> ignore)) |> ignore
        Assert.Throws<ArgumentException>(Action(fun () ->
            ProviderTrustOfflinePolicy.create (TimeSpan.FromHours 25.0) (TimeSpan.FromMinutes 1.0) |> ignore)) |> ignore
        let result = (policy()).Evaluate(
            start.AddMinutes 2.0, Some start, "policy-poll-exhausted", Some "policy-distribution-timeout", 1)
        let overflowing = (policy()).Evaluate(
            DateTimeOffset.MaxValue, Some(DateTimeOffset.MaxValue.AddMinutes -1.0),
            "policy-poll-exhausted", Some "policy-distribution-timeout", 1)
        Assert.Multiple(Action(fun () ->
            Assert.That(result.Deadline, Is.EqualTo(Some(start.AddMinutes 5.0)))
            Assert.That(result.RetryAt, Is.EqualTo(Some(start.AddMinutes 3.0)))
            Assert.That(overflowing.Code, Is.EqualTo "offline-observation-invalid")))

    [<Test>]
    member _.``CBI49 C2 only endpoint unavailability is grace eligible``() =
        let cases =
            [ "policy-poll-refused", "policy-distribution-endpoint-signature-invalid"
              "policy-poll-exhausted", "policy-distribution-stale"
              "policy-poll-exhausted", "policy-distribution-superseded"
              "policy-poll-canceled", "policy-distribution-canceled"
              "policy-poll-floor-unretained", "policy-distribution-update-applied" ]
        for pollCode, attemptCode in cases do
            let result = (policy()).Evaluate(start.AddMinutes 1.0, Some start, pollCode, Some attemptCode, 1)
            Assert.That(result.Code, Is.EqualTo "offline-service-stop-required", $"{pollCode}/{attemptCode}")
            Assert.That(result.MayContinueExistingService, Is.False)

    [<Test>]
    member _.``CBI49 C3 grace requires prior current and never refreshes it``() =
        let subject = policy ()
        let evaluate now baseline =
            subject.Evaluate(now, baseline, "policy-poll-exhausted", Some "policy-distribution-timeout", 1)
        let noBaseline = evaluate (start.AddMinutes 1.0) None
        let before = evaluate (start.AddMinutes 4.0) (Some start)
        let atDeadline = evaluate (start.AddMinutes 5.0) (Some start)
        let later = evaluate (start.AddMinutes 6.0) (Some start)
        Assert.Multiple(Action(fun () ->
            Assert.That(noBaseline.Code, Is.EqualTo "offline-service-stop-required")
            Assert.That(before.Code, Is.EqualTo "offline-existing-service")
            Assert.That(atDeadline.Code, Is.EqualTo "offline-grace-expired")
            Assert.That(later.Code, Is.EqualTo "offline-grace-expired")
            Assert.That(later.Deadline, Is.EqualTo before.Deadline)))

    [<Test>]
    member _.``CBI49 C4 offline continuation is existing service only``() =
        let subject = policy ()
        let evaluate count =
            subject.Evaluate(start.AddMinutes 2.0, Some start,
                "policy-poll-exhausted", Some "policy-distribution-transport-failed", count)
        let serving = evaluate 2
        let idle = evaluate 0
        Assert.Multiple(Action(fun () ->
            Assert.That(serving.MayContinueExistingService, Is.True)
            Assert.That(serving.MayStartProvider, Is.False)
            Assert.That(idle.Code, Is.EqualTo "offline-idle")
            Assert.That(idle.MayContinueExistingService, Is.False)
            Assert.That(idle.MayStartProvider, Is.False)))

    [<Test>]
    member _.``CBI49 C5 reconciliation evidence names the interrupted attempt exactly``() =
        use temporary = new Cbi49TemporaryJournal()
        let journal = interrupted temporary.Path
        let before = File.ReadAllBytes temporary.Path
        let evidence =
            { RunIdentity = runIdentity; AttemptIndex = 1; AttemptInstant = start
              Verdict = ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed }
        let result = ProviderTrustCadenceReconciliation.apply journal evidence
        Assert.Multiple(Action(fun () ->
            Assert.That(result.Code, Is.EqualTo "cadence-reconciliation-mismatch")
            Assert.That(result.Snapshot.Phase, Is.EqualTo "in-flight")
            Assert.That(Convert.ToHexString(File.ReadAllBytes temporary.Path), Is.EqualTo(Convert.ToHexString before))))

    [<Test>]
    member _.``CBI49 C6 unknown evidence leaves the interruption inert``() =
        use temporary = new Cbi49TemporaryJournal()
        let journal = interrupted temporary.Path
        let before = File.ReadAllBytes temporary.Path
        let evidence =
            { RunIdentity = runIdentity; AttemptIndex = 0; AttemptInstant = start
              Verdict = ProviderTrustCadenceReconciliationVerdict.Unknown }
        let result = ProviderTrustCadenceReconciliation.apply journal evidence
        Assert.Multiple(Action(fun () ->
            Assert.That(result.Code, Is.EqualTo "cadence-reconciliation-deferred")
            Assert.That(result.Snapshot.Phase, Is.EqualTo "in-flight")
            Assert.That(result.Snapshot.InterruptionCount, Is.Zero)
            Assert.That(Convert.ToHexString(File.ReadAllBytes temporary.Path), Is.EqualTo(Convert.ToHexString before))))

    [<Test>]
    member _.``CBI49 C7 conclusive evidence selects one CBI48 transition``() =
        use retryTemporary = new Cbi49TemporaryJournal()
        use abandonTemporary = new Cbi49TemporaryJournal()
        let retry = interrupted retryTemporary.Path
        let abandon = interrupted abandonTemporary.Path
        let evidence verdict =
            { RunIdentity = runIdentity; AttemptIndex = 0; AttemptInstant = start; Verdict = verdict }
        let retried = ProviderTrustCadenceReconciliation.apply retry (evidence ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed)
        let repeated = ProviderTrustCadenceReconciliation.apply retry (evidence ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed)
        let abandoned = ProviderTrustCadenceReconciliation.apply abandon (evidence ProviderTrustCadenceReconciliationVerdict.EffectsAccountedFor)
        Assert.Multiple(Action(fun () ->
            Assert.That(retried.Code, Is.EqualTo "cadence-reconciliation-retry-ready")
            Assert.That(retried.Snapshot.InterruptionCount, Is.EqualTo 1)
            Assert.That(retried.Snapshot.RetryCount, Is.EqualTo 1)
            Assert.That(repeated.Code, Is.EqualTo "cadence-reconciliation-not-required")
            Assert.That(repeated.Snapshot.InterruptionCount, Is.EqualTo 1)
            Assert.That(abandoned.Code, Is.EqualTo "cadence-reconciliation-abandoned")
            Assert.That(abandoned.Snapshot.InterruptionCount, Is.EqualTo 1)
            Assert.That(abandoned.Snapshot.RetryCount, Is.Zero)))

    [<Test>]
    member _.``CBI49 C8 minimal executes the shared policy model``() =
        let fixturePath = Path.Combine(
            TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
            "cbi49-offline-reconciliation-vectors.json")
        use document = JsonDocument.Parse(File.ReadAllText fixturePath)
        let root = document.RootElement
        let fixturePolicy = root.GetProperty "policy"
        let graceSeconds = (fixturePolicy.GetProperty "graceSeconds").GetDouble()
        let retrySeconds = (fixturePolicy.GetProperty "retrySeconds").GetDouble()
        let subject = ProviderTrustOfflinePolicy.create (TimeSpan.FromSeconds graceSeconds) (TimeSpan.FromSeconds retrySeconds)
        let fixtureLastCurrent = (fixturePolicy.GetProperty "lastCurrent").GetDateTimeOffset()
        let stringProperty (element: JsonElement) (name: string) : string =
            match (element.GetProperty name).GetString() with
            | null -> failwith $"Fixture property {name} is null."
            | value -> value
        for vector in (root.GetProperty "offlineVectors").EnumerateArray() do
            let id = stringProperty vector "id"
            let baseline =
                if id = "no-current-baseline" then None
                else
                    let mutable property = Unchecked.defaultof<JsonElement>
                    if vector.TryGetProperty("lastCurrent", &property) && property.ValueKind <> JsonValueKind.Null then
                        Some(property.GetDateTimeOffset())
                    else Some fixtureLastCurrent
            let lastAttempt =
                let mutable property = Unchecked.defaultof<JsonElement>
                if vector.TryGetProperty("lastAttemptCode", &property) && property.ValueKind <> JsonValueKind.Null then
                    property.GetString() |> Option.ofObj
                else None
            let result = subject.Evaluate(
                vector.GetProperty("now").GetDateTimeOffset(), baseline,
                stringProperty vector "pollCode", lastAttempt,
                vector.GetProperty("servingCount").GetInt32())
            let optionalInstant (name: string) =
                let property = vector.GetProperty name
                if property.ValueKind = JsonValueKind.Null then None else Some(property.GetDateTimeOffset())
            Assert.Multiple(Action(fun () ->
                Assert.That(result.Code, Is.EqualTo(stringProperty vector "expectedCode"), id)
                Assert.That(result.MayContinueExistingService,
                    Is.EqualTo(vector.GetProperty("continueExisting").GetBoolean()), id)
                Assert.That(result.MayStartProvider, Is.EqualTo(vector.GetProperty("mayStart").GetBoolean()), id)
                Assert.That(result.Deadline, Is.EqualTo(optionalInstant "deadline"), id)
                Assert.That(result.RetryAt, Is.EqualTo(optionalInstant "retryAt"), id)))

        for vector in (root.GetProperty "reconciliationVectors").EnumerateArray() do
            use temporary = new Cbi49TemporaryJournal()
            let journal = interrupted temporary.Path
            let verdict =
                match stringProperty vector "verdict" with
                | "no-effects-confirmed" -> ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed
                | "effects-accounted-for" -> ProviderTrustCadenceReconciliationVerdict.EffectsAccountedFor
                | _ -> ProviderTrustCadenceReconciliationVerdict.Unknown
            let mutable indexProperty = Unchecked.defaultof<JsonElement>
            let index =
                if vector.TryGetProperty("attemptIndex", &indexProperty) then indexProperty.GetInt32() else 0
            let evidence =
                { RunIdentity = runIdentity; AttemptIndex = index
                  AttemptInstant = start; Verdict = verdict }
            let result = ProviderTrustCadenceReconciliation.apply journal evidence
            let id = stringProperty vector "id"
            Assert.Multiple(Action(fun () ->
                Assert.That(result.Code, Is.EqualTo(stringProperty vector "expectedCode"), id)
                Assert.That(result.Snapshot.Phase, Is.EqualTo(stringProperty vector "expectedPhase"), id)
                Assert.That(result.Snapshot.InterruptionCount,
                    Is.EqualTo(vector.GetProperty("interruptions").GetInt32()), id)
                Assert.That(result.Snapshot.RetryCount, Is.EqualTo(vector.GetProperty("retries").GetInt32()), id)))

File: Reference/tests/Brontide.Reference.Studio.Tests/ComponentProviderTrustOfflinePolicyTests.cs
using System.Text.Json;

namespace Brontide.Reference.Studio.Tests;

public sealed class ComponentProviderTrustOfflinePolicyTests
{
    private sealed class TemporaryJournal : IDisposable
    {
        private readonly string root = System.IO.Path.Combine(
            System.IO.Path.GetTempPath(), $"brontide-cbi49-{Guid.NewGuid():N}");

        public string Path => System.IO.Path.Combine(root, "cadence.bin");

        public void Dispose()
        {
            try { if (Directory.Exists(root)) Directory.Delete(root, true); }
            catch (IOException) { }
            catch (UnauthorizedAccessException) { }
        }
    }

    private sealed record Fixture(
        FixturePolicy Policy,
        IReadOnlyList<OfflineVector> OfflineVectors,
        IReadOnlyList<ReconciliationVector> ReconciliationVectors);

    private sealed record FixturePolicy(int GraceSeconds, int RetrySeconds, DateTimeOffset LastCurrent);

    private sealed record OfflineVector(
        string Id,
        DateTimeOffset Now,
        DateTimeOffset? LastCurrent,
        string PollCode,
        string? LastAttemptCode,
        int ServingCount,
        string ExpectedCode,
        bool ContinueExisting,
        bool MayStart,
        DateTimeOffset? Deadline,
        DateTimeOffset? RetryAt);

    private sealed record ReconciliationVector(
        string Id,
        string Verdict,
        int? AttemptIndex,
        string ExpectedCode,
        string ExpectedPhase,
        int Interruptions,
        int Retries);

    private static readonly ProviderTrustCadenceRunId RunId =
        ProviderTrustCadenceRunId.Create("cadence-run.cbi49");

    private static readonly DateTimeOffset Start =
        new(2026, 8, 5, 8, 0, 0, TimeSpan.Zero);

    private static readonly ProviderServingTrustCadenceSchedule Schedule =
        ProviderServingTrustCadenceSchedule.Create(2, TimeSpan.FromMinutes(1));

    private static Fixture LoadFixture()
    {
        var path = Path.Combine(TestContext.CurrentContext.TestDirectory,
            "component-management", "fixtures", "cbi49-offline-reconciliation-vectors.json");
        return JsonSerializer.Deserialize<Fixture>(File.ReadAllText(path), new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true,
        })!;
    }

    private static DurableProviderTrustCadenceJournal Interrupted(string path)
    {
        var journal = DurableProviderTrustCadenceJournal.Establish(path, RunId, Schedule, Start).Journal!;
        Assert.That(journal.BeginCycle().Code, Is.EqualTo("durable-cadence-cycle-started"));
        return DurableProviderTrustCadenceJournal.Open(path, RunId).Journal!;
    }

    [Test]
    public void Cbi49_C1_offline_policy_is_explicit_and_bounded()
    {
        Assert.Multiple(() =>
        {
            Assert.Throws<ArgumentOutOfRangeException>(() =>
                ProviderTrustOfflinePolicy.Create(TimeSpan.Zero, TimeSpan.FromMinutes(1)));
            Assert.Throws<ArgumentOutOfRangeException>(() =>
                ProviderTrustOfflinePolicy.Create(TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(2)));
            Assert.Throws<ArgumentOutOfRangeException>(() =>
                ProviderTrustOfflinePolicy.Create(TimeSpan.FromHours(25), TimeSpan.FromMinutes(1)));
        });

        var policy = ProviderTrustOfflinePolicy.Create(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1));
        var result = policy.Evaluate(Start.AddMinutes(2), Start,
            "policy-poll-exhausted", "policy-distribution-timeout", 1);
        var overflowing = policy.Evaluate(DateTimeOffset.MaxValue,
            DateTimeOffset.MaxValue.AddMinutes(-1),
            "policy-poll-exhausted", "policy-distribution-timeout", 1);
        Assert.Multiple(() =>
        {
            Assert.That(result.Deadline, Is.EqualTo(Start.AddMinutes(5)));
            Assert.That(result.RetryAt, Is.EqualTo(Start.AddMinutes(3)));
            Assert.That(overflowing.Code, Is.EqualTo("offline-observation-invalid"));
        });
    }

    [Test]
    public void Cbi49_C2_only_endpoint_unavailability_is_grace_eligible()
    {
        var policy = ProviderTrustOfflinePolicy.Create(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1));
        foreach (var (poll, attempt) in new[]
        {
            ("policy-poll-refused", "policy-distribution-endpoint-signature-invalid"),
            ("policy-poll-exhausted", "policy-distribution-stale"),
            ("policy-poll-exhausted", "policy-distribution-superseded"),
            ("policy-poll-canceled", "policy-distribution-canceled"),
            ("policy-poll-floor-unretained", "policy-distribution-update-applied"),
        })
        {
            var result = policy.Evaluate(Start.AddMinutes(1), Start, poll, attempt, 1);
            Assert.That(result.Code, Is.EqualTo("offline-service-stop-required"), $"{poll}/{attempt}");
            Assert.That(result.MayContinueExistingService, Is.False);
        }
    }

    [Test]
    public void Cbi49_C3_grace_requires_prior_current_and_never_refreshes_it()
    {
        var policy = ProviderTrustOfflinePolicy.Create(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1));
        var noBaseline = policy.Evaluate(Start.AddMinutes(1), null,
            "policy-poll-exhausted", "policy-distribution-timeout", 1);
        var before = policy.Evaluate(Start.AddMinutes(4), Start,
            "policy-poll-exhausted", "policy-distribution-timeout", 1);
        var atDeadline = policy.Evaluate(Start.AddMinutes(5), Start,
            "policy-poll-exhausted", "policy-distribution-timeout", 1);
        var later = policy.Evaluate(Start.AddMinutes(6), Start,
            "policy-poll-exhausted", "policy-distribution-timeout", 1);

        Assert.Multiple(() =>
        {
            Assert.That(noBaseline.Code, Is.EqualTo("offline-service-stop-required"));
            Assert.That(before.Code, Is.EqualTo("offline-existing-service"));
            Assert.That(atDeadline.Code, Is.EqualTo("offline-grace-expired"));
            Assert.That(later.Code, Is.EqualTo("offline-grace-expired"));
            Assert.That(later.Deadline, Is.EqualTo(before.Deadline));
        });
    }

    [Test]
    public void Cbi49_C4_offline_continuation_is_existing_service_only()
    {
        var policy = ProviderTrustOfflinePolicy.Create(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1));
        var serving = policy.Evaluate(Start.AddMinutes(2), Start,
            "policy-poll-exhausted", "policy-distribution-transport-failed", 2);
        var idle = policy.Evaluate(Start.AddMinutes(2), Start,
            "policy-poll-exhausted", "policy-distribution-transport-failed", 0);

        Assert.Multiple(() =>
        {
            Assert.That(serving.MayContinueExistingService, Is.True);
            Assert.That(serving.MayStartProvider, Is.False);
            Assert.That(idle.Code, Is.EqualTo("offline-idle"));
            Assert.That(idle.MayContinueExistingService, Is.False);
            Assert.That(idle.MayStartProvider, Is.False);
        });
    }

    [Test]
    public void Cbi49_C5_reconciliation_evidence_names_the_interrupted_attempt_exactly()
    {
        using var temporary = new TemporaryJournal();
        var journal = Interrupted(temporary.Path);
        var before = File.ReadAllBytes(temporary.Path);
        var mismatch = new ProviderTrustCadenceReconciliationEvidence(
            RunId, 1, Start, ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed);
        var result = ProviderTrustCadenceReconciliation.Apply(journal, mismatch);

        Assert.Multiple(() =>
        {
            Assert.That(result.Code, Is.EqualTo("cadence-reconciliation-mismatch"));
            Assert.That(result.Snapshot.Phase, Is.EqualTo("in-flight"));
            Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(before));
        });
    }

    [Test]
    public void Cbi49_C6_unknown_evidence_leaves_the_interruption_inert()
    {
        using var temporary = new TemporaryJournal();
        var journal = Interrupted(temporary.Path);
        var before = File.ReadAllBytes(temporary.Path);
        var evidence = new ProviderTrustCadenceReconciliationEvidence(
            RunId, 0, Start, ProviderTrustCadenceReconciliationVerdict.Unknown);
        var result = ProviderTrustCadenceReconciliation.Apply(journal, evidence);

        Assert.Multiple(() =>
        {
            Assert.That(result.Code, Is.EqualTo("cadence-reconciliation-deferred"));
            Assert.That(result.Snapshot.Phase, Is.EqualTo("in-flight"));
            Assert.That(result.Snapshot.InterruptionCount, Is.Zero);
            Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(before));
        });
    }

    [Test]
    public void Cbi49_C7_conclusive_evidence_selects_one_cbi48_transition()
    {
        using var retryTemporary = new TemporaryJournal();
        using var abandonTemporary = new TemporaryJournal();
        var retry = Interrupted(retryTemporary.Path);
        var abandon = Interrupted(abandonTemporary.Path);

        var retried = ProviderTrustCadenceReconciliation.Apply(retry,
            new(RunId, 0, Start, ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed));
        var repeated = ProviderTrustCadenceReconciliation.Apply(retry,
            new(RunId, 0, Start, ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed));
        var abandoned = ProviderTrustCadenceReconciliation.Apply(abandon,
            new(RunId, 0, Start, ProviderTrustCadenceReconciliationVerdict.EffectsAccountedFor));

        Assert.Multiple(() =>
        {
            Assert.That(retried.Code, Is.EqualTo("cadence-reconciliation-retry-ready"));
            Assert.That(retried.Snapshot.InterruptionCount, Is.EqualTo(1));
            Assert.That(retried.Snapshot.RetryCount, Is.EqualTo(1));
            Assert.That(repeated.Code, Is.EqualTo("cadence-reconciliation-not-required"));
            Assert.That(repeated.Snapshot.InterruptionCount, Is.EqualTo(1));
            Assert.That(abandoned.Code, Is.EqualTo("cadence-reconciliation-abandoned"));
            Assert.That(abandoned.Snapshot.InterruptionCount, Is.EqualTo(1));
            Assert.That(abandoned.Snapshot.RetryCount, Is.Zero);
        });
    }

    [Test]
    public void Cbi49_C8_reference_executes_the_shared_policy_model()
    {
        var fixture = LoadFixture();
        var policy = ProviderTrustOfflinePolicy.Create(
            TimeSpan.FromSeconds(fixture.Policy.GraceSeconds),
            TimeSpan.FromSeconds(fixture.Policy.RetrySeconds));

        foreach (var vector in fixture.OfflineVectors)
        {
            DateTimeOffset? baseline = vector.Id == "no-current-baseline"
                ? null
                : vector.LastCurrent ?? fixture.Policy.LastCurrent;
            var result = policy.Evaluate(vector.Now, baseline,
                vector.PollCode, vector.LastAttemptCode, vector.ServingCount);
            Assert.Multiple(() =>
            {
                Assert.That(result.Code, Is.EqualTo(vector.ExpectedCode), vector.Id);
                Assert.That(result.MayContinueExistingService, Is.EqualTo(vector.ContinueExisting), vector.Id);
                Assert.That(result.MayStartProvider, Is.EqualTo(vector.MayStart), vector.Id);
                Assert.That(result.Deadline, Is.EqualTo(vector.Deadline), vector.Id);
                Assert.That(result.RetryAt, Is.EqualTo(vector.RetryAt), vector.Id);
            });
        }

        foreach (var vector in fixture.ReconciliationVectors)
        {
            using var temporary = new TemporaryJournal();
            var journal = Interrupted(temporary.Path);
            var verdict = vector.Verdict switch
            {
                "no-effects-confirmed" => ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed,
                "effects-accounted-for" => ProviderTrustCadenceReconciliationVerdict.EffectsAccountedFor,
                _ => ProviderTrustCadenceReconciliationVerdict.Unknown,
            };
            var result = ProviderTrustCadenceReconciliation.Apply(journal,
                new(RunId, vector.AttemptIndex ?? 0, Start, verdict));
            Assert.Multiple(() =>
            {
                Assert.That(result.Code, Is.EqualTo(vector.ExpectedCode), vector.Id);
                Assert.That(result.Snapshot.Phase, Is.EqualTo(vector.ExpectedPhase), vector.Id);
                Assert.That(result.Snapshot.InterruptionCount, Is.EqualTo(vector.Interruptions), vector.Id);
                Assert.That(result.Snapshot.RetryCount, Is.EqualTo(vector.Retries), vector.Id);
            });
        }
    }
}
